Adopt smart pointers in ApplicationStateTracker

brentfulgham opened the pull request:
#### 6c333c99a9a77e7f985b532002f6f2f8f7e05a07
<pre>
Adopt smart pointers in ApplicationStateTracker
<a href="https://bugs.webkit.org/show_bug.cgi?id=273218">https://bugs.webkit.org/show_bug.cgi?id=273218</a>
&lt;<a href="https://rdar.apple.com/126764684">rdar://126764684</a>&gt;

Reviewed by David Kilzer.

The ApplicationStateTracker uses `id` values to remember observers that it
Registers and unregisters at various points in the display of a view. It&apos;s
possible that some framework could decide to unregistered an observer outside
of this flow, causing us to request an observer to be removed a second time,
which triggers a release assert.

* Source/WebKit/UIProcess/ApplicationStateTracker.h:
* Source/WebKit/UIProcess/ApplicationStateTracker.mm:
(WebKit::m_isInBackground):
(WebKit::ApplicationStateTracker::removeAllObservers):
(WebKit::m_willEnterForegroundObserver): Deleted.

Canonical link: <a href="https://commits.webkit.org/278022@main">https://commits.webkit.org/278022@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9355f1a551bc60f7888ee16ceffd66faace25007

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49185 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28469 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52212 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51982 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/45285 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34467 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26068 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40194 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51129 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26027 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42389 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21309 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23482 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7508 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45406 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44071 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53893 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24267 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20468 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25540 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42593 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10830 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26371 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25261 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebKit/UIProcess/ApplicationStateTracker.h:
```diff
@@ -83,10 +83,10 @@ class ApplicationStateTracker : public CanMakeWeakPtr<ApplicationStateTracker> {
+    RetainPtr<NSObject> m_didEnterBackgroundObserver;
+    RetainPtr<NSObject> m_willEnterForegroundObserver;
+    RetainPtr<NSObject> m_willBeginSnapshotSequenceObserver;
+    RetainPtr<NSObject> m_didCompleteSnapshotSequenceObserver;
```
Generally speaking, it's not a good idea to retain observer objects because that has the potential to create a retain cycle with `this` (C++) or `self` (Objective-C).

Also, if the client (third-party) software wants to deallocate the observer object sooner, this object keeping them alive may cause other issues.

So I would suggest using `WeakObjCPtr<>` (see lines 70-73 above) or adding `__weak` to these instance variables so that you keep a weak reference to them.

If you use `WeakObjCPtr<>`, then the rest of the patch doesn't need to change.

If you use `__weak`, then you just need to get rid of the `.get()` calls below.

ddkilzer replied:
I thought about this some more, and I think `RetainPtr<>` is okay if `ApplicationStateTracker` is not a (super) long-lived.

The reason objects like `NSNotificationCenter` don't take strong references is that if registered observers aren't removed, they'll basically leak for the life of the process.

So using either a weak or a strong reference is fine, though I'd still lean towards using a weak reference if possible (since it has the same behavior as a raw pointer, but without the downside of stale pointers when the observer object is deallocated but not unregistered).

brentfulgham replied:
I'm testing with the `WeakObjCPtr<>` approach, and will very likely land that version.

ddkilzer reviewed (approved):
r=me if you hold the observer objects weakly instead of strongly.

brentfulgham committed:
Adopt smart pointers in ApplicationStateTracker
https://bugs.webkit.org/show_bug.cgi?id=273218
<rdar://126764684>

Reviewed by David Kilzer.

The ApplicationStateTracker uses `id` values to remember observers that it
Registers and unregisters at various points in the display of a view. It's
possible that some framework could decide to unregistered an observer outside
of this flow, causing us to request an observer to be removed a second time,
which triggers a release assert.

* Source/WebKit/UIProcess/ApplicationStateTracker.h:
* Source/WebKit/UIProcess/ApplicationStateTracker.mm:
(WebKit::m_isInBackground):
(WebKit::ApplicationStateTracker::removeAllObservers):
(WebKit::m_willEnterForegroundObserver): Deleted.

Canonical link: https://commits.webkit.org/278022@main
